torch/nn/modules/module.py

```diff
@@ -45,14 +45,30 @@ def _addindent(s_, numSpaces):
     s = first + '\n' + s
     return s
 
+class SharedBool(object):
+    def __init__(self, val):
+        self.value = val
+
+
+class CachedOrderedDict(OrderedDict):
+    def __init__(self, shared_bool: SharedBool):
+        super().__init__()
+        self.is_updated = shared_bool
+
+    def __setitem__(self, key, value):
+        self.is_updated.value = True
```

Comment:
How reliable is this kind of trick? I'm sure there are ways to modify the dict without going through this function?

Reply:
```
from collections import OrderedDict

class SentinelDict(OrderedDict):
    def __setitem__(self, key, value):
      print(f"__setitem__  {key}: {value}")
      super().__setitem__(key, value)

x = SentinelDict()

# Touches __setitem__
x["a"] = 1
x.setdefault("b", 2)

# Mutates without __setitem__
x.pop("b")
x.clear()
```

Reply:
Unfortunately, in Python many of these things are near impossible to guarantee. If you have a suggestion for how to do that reliably I would love that.

Reply:
Ho that's not great :/
While I agree that we should not optimize for people doing super adversarial things, I do think that correctness is more important than anything else here.
I don't think our hook mechanism uses `pop()` right now to remove the handles. But that could be very much possible (and I would expect that to work).

Maybe we should consider maintaining this flag directly from the `register_XXX_hook` functions so that people that use the right API can never get a bad result?